File: wire-schema/src/main/java/com/squareup/wire/schema/Extensions.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema;

import com.google.common.collect.ImmutableList;
import com.squareup.wire.schema.internal.Util;
import com.squareup.wire.schema.internal.parser.ExtensionsElement;

final class Extensions {
  private final Location location;
  private final String documentation;
  private final int start;
  private final int end;

  private Extensions(Location location, String documentation, int start, int end) {
    this.location = location;
    this.documentation = documentation;
    this.start = start;
    this.end = end;
  }

  public Location location() {
    return location;
  }

  public String documentation() {
    return documentation;
  }

  public int start() {
    return start;
  }

  public int end() {
    return end;
  }

  static ImmutableList<Extensions> fromElements(ImmutableList<ExtensionsElement> elements) {
    ImmutableList.Builder<Extensions> extensions = ImmutableList.builder();
    for (ExtensionsElement element : elements) {
      extensions.add(new Extensions(element.location(), element.documentation(),
          element.start(), element.end()));
    }
    return extensions.build();
  }

  static ImmutableList<ExtensionsElement> toElements(ImmutableList<Extensions> extensions) {
    ImmutableList.Builder<ExtensionsElement> elements = new ImmutableList.Builder<>();
    for (Extensions extension : extensions) {
      elements.add(ExtensionsElement.create(extension.location, extension.start, extension.end,
          extension.documentation));
    }
    return elements.build();
  }

  void validate(Linker linker) {
    if (!Util.isValidTag(start()) || !Util.isValidTag(end())) {
      linker.withContext(this).addError("tags are out of range: %s to %s", start(), end());
    }
  }
}

File: wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Extensions.kt
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema

import com.squareup.wire.schema.internal.isValidTag
import com.squareup.wire.schema.internal.parser.ExtensionsElement
import kotlin.jvm.JvmStatic

data class Extensions(
  val location: Location,
  val documentation: String,
  val values: List<Any>,
) {
  fun validate(linker: Linker) {
    @Suppress("NAME_SHADOWING")
    val linker = linker.withContext(this)
    val outOfRangeTags = mutableListOf<String>()
    values.forEach { value ->
      when (value) {
        is Int -> {
          if (!value.isValidTag()) {
            outOfRangeTags.add("$value")
          }
        }
        is IntRange -> {
          if (!value.first.isValidTag() || !value.last.isValidTag()) {
            outOfRangeTags.add("${value.first} to ${value.last}")
          }
        }
        else -> throw AssertionError()
      }
    }
    if (outOfRangeTags.isNotEmpty()) {
      linker.errors += "tags are out of range: ${outOfRangeTags.joinToString(separator = ", ")}"
    }
  }

  companion object {
    @JvmStatic
    fun fromElements(elements: List<ExtensionsElement>) =
      elements.map { Extensions(it.location, it.documentation, it.values) }

    @JvmStatic
    fun toElements(extensions: List<Extensions>) =
      extensions.map { ExtensionsElement(it.location, it.documentation, it.values) }
  }
}
